### PR TITLE
[Fix] Correct the final-state gradient for chunks shorter than `W - 1`

### LINKED ISSUE #1140: [Bug] causal_conv1d: wrong gradients when a chunk is shorter than W - 1 tokens
### Checklist

- [x] I have checked [FAQs](https://github.com/fla-org/flash-linear-attention/blob/main/FAQs.md) and existing issues for similar problems
- [x] Please report this bug in English to ensure wider understanding and support

### Describe the Bug

The backward routes the final-state gradient `dht` back to the input tokens that produced it. `final_state[d, w]` is input token `t = T - W + w`, so token `t` must receive `dht[d, t + W - T]`. `causal_conv1d_bwd_kernel` derives that slot index from `start_tok = max(0, T - (W - 1))`, and the `max(0, ...)` is the bug: as soon as `T < W - 1` it pins `start_tok` at 0, and every contribution is shifted by `W - 1 - T` slots. For `T >= W - 1` the clamp does nothing and the kernel is correct. fla's own hand-written triton-ascend port of this kernel has the unclamped expression.

A sequence that short also leaves part of `initial_state` still sitting in `final_state` — slot `i_w` survives into `final_state[:, i_w - T]` whenever `i_w >= T` — and `compute_dh0_kernel` drops that term, so `dh0` is wrong under the same condition; the two together are what makes a short chunk's gradients disagree. With the shipped conv width `W = 4` this means chunks of 1 or 2 tokens — single-token decode and short chunked prefill, where the first chunk's `final_state` is handed to the next chunk as its `initial_state`. The forward is exact in these cases, only the gradients are wrong, nothing raises, and the existing backward tests all use `T >= 15`, where the clamp is inert.

https://github.com/fla-org/flash-linear-attention/blob/033b19e81239b13971a410e55dd6c178d430b9d4/fla/modules/conv/triton/kernels.py#L317-L327

The final-state gradient path was added in #568.

### Steps to Reproduce the Bug

fla's own composition contract: running a sequence in two chunks, feeding the first chunk's `final_state` to the second as `initial_state`, must give the same gradients as running the whole sequence at once.

```python
import torch
from fla.modules.conv import causal_conv1d

device, B, T, D, W = 'cuda', 1, 16, 16, 4

def chunked(T1):
    torch.manual_seed(42)
    x = torch.randn(B, T, D, device=device, requires_grad=True)
    weight = torch.randn(D, W, device=device, requires_grad=True)
    bias = torch.randn(D, device=device, requires_grad=True)
    h0 = torch.randn(B, D, W, device=device, requires_grad=True)
    dy = torch.randn(B, T, D, device=device)

    y, _ = causal_conv1d(x, weight, bias, initial_state=h0, activation='silu')
    one_shot = torch.autograd.grad(y, [x, h0], grad_outputs=dy, retain_graph=True)

    y1, state = causal_conv1d(x[:, :T1], weight, bias, initial_state=h0,
                              output_final_state=True, activation='silu')
    y2, _ = causal_conv1d(x[:, T1:], weight, bias, initial_state=state, activation='silu')
    y_chunked = torch.cat([y1, y2], dim=1)
    split = torch.autograd.grad(y_chunked, [x, h0], grad_outputs=dy)

    return ((y_chunked - y).abs().max().item(),
            [(a - b).abs().max().item() for a, b in zip(split, one_shot)])

print(f'{"first chunk":>12} {"forward":>10} {"dx":>10} {"dh0":>10}')
for T1 in [1, 2, 3, 4, 8]:
    fwd, errs = chunked(T1)
    print(f'{T1:>9} tok {fwd:10.2e} {errs[0]:10.4f} {errs[1]:10.4f}')
```

```
 first chunk    forward         dx        dh0
        1 tok   0.00e+00     2.4404     1.3728
        2 tok   0.00e+00     4.4926     1.3728
        3 tok   0.00e+00     0.0000     0.0000
        4 tok   0.00e+00     0.0000     0.0000
        8 tok   0.00e+00     0.0000     0.0000
```

The one-shot call itself matches a plain torch autograd reference to `< 1e-4` at every split, so the disagreement is entirely in the short first chunk's backward, and the break is exactly at `T1 < W - 1 = 3`.

### Expected Behavior

The same gradients whatever the chunk boundaries, as the forward already gives: the slot index taken from the unclamped `T - (W - 1)` (the `(offset >= start_tok) & (offset < T)` mask already keeps the reads in range, and the index stays inside `[1, W)` for every `T >= 1`), and `dh0` also picking up the `dht` slots the chunk was too short to overwrite.

### Environment Information

1. Torch: 2.13.0+cu130
2. Triton: 3.7.1
3. Commit tested: 033b19e8 (current main), Python 3.12, NVIDIA B200, driver CUDA 13.2

### PR DESCRIPTION
**Pull request by truong-v**
## Summary

`final_state[d, w]` is input token `t = T - W + w`, so token `t` must receive `dht[d, t + W - T]`; `causal_conv1d_bwd_kernel` derives that slot index from `start_tok = max(0, T - (W - 1))`, and the clamp — inert for `T >= W - 1` — pins `start_tok` at 0 below it and shifts every contribution by `W - 1 - T` slots. A sequence that short also leaves part of `initial_state` still sitting in `final_state` (slot `i_w` survives into `final_state[:, i_w - T]` whenever `i_w >= T`), a term `compute_dh0_kernel` dropped, so at the shipped conv width `W = 4` a 1- or 2-token prefill chunk disagrees with the same sequence run in one call while the forward stays exact and nothing raises.

Fixes #1140.

- The `dht` slot index comes from the unclamped `offset + (W - T)`. `start_tok` keeps its clamp for the mask, where it is equivalent (`offset >= 0` regardless), so the diff touches the slot index only.
- `compute_dh0_kernel` takes `dht` and adds the pass-through term for the slots a short sequence does not overwrite.

## Test plan

- `python scripts/find_dependent_tests.py fla/modules/conv/triton/kernels.py` → `tests/modules/test_conv.py` (plus the layer/model tests that build on it). `tests/modules/test_conv.py` on a B200: 84 passed, 26 skipped — 77 passed / 26 skipped at unmodified main, plus the 7 new cases.
- New `test_conv_chunked_prefill_backward`: runs a sequence in two chunks, handing the first chunk's `final_state` to the second as `initial_state`, and requires the gradients to match the one-shot call over the whole sequence — the composition the forward already satisfies exactly. Parametrized over first chunks of 1, 2, 3 and 8 tokens, so it covers both sides of the `T = W - 1` boundary, plus a `W = 3` case. Reverting the kernel change alone fails all five 1- and 2-token cases and leaves the two 3- and 8-token cases passing.
- The existing suite could not see this: every backward test uses `T >= 15`, well above `W - 1`, where the clamp is inert and a chunk always overwrites the whole state.
- Max abs error of the two-chunk gradients against the one-shot call at B=1, T=16, D=16, W=4, `silu`:

| first chunk | dx before | dx after | dh0 before | dh0 after |
|---|---|---|---|---|
| 1 token | 2.4404 | 0.0000 | 1.3728 | 0.0000 |
| 2 tokens | 4.4926 | 0.0000 | 1.3728 | 0.0000 |
| 3 tokens | 0.0000 | 0.0000 | 0.0000 | 0.0000 |
| 8 tokens | 0.0000 | 0.0000 | 0.0000 | 0.0000 |

## Benchmark / NCU

`triton.testing.do_bench` on an NVIDIA B200, bf16, `W = 4`, `activation='silu'`. Autotune paid outside the timed region, min of 7 `do_bench` calls, three interleaved before/after rounds. Varlen is `[1, 16384, 2048]` with `cu_seqlens=[0, 3000, 7001, 11002, 16384]`; the short-chunk shape is `[64, 2, 2048]` — the case the new `dht` term actually serves.

| workload | before | after |
|---|---|---|
| `causal_conv1d_bwd`, varlen, `dht` only (isolates the slot-index change) | 318.1 / 317.9 / 317.8 us | 317.6 / 320.5 / 317.8 us |
| `causal_conv1d_bwd`, varlen, initial_state + `dht` | 485.8 / 485.7 / 485.7 us | 487.2 / 489.1 / 486.9 us |
| `causal_conv1d_bwd`, dense, T=2, B=64 | 170.0 / 169.8 / 169.8 us | 170.4 / 170.7 / 170.6 us |
| `compute_dh0_triton`, dense, T=2, B=64 | 11.9 / 11.9 / 11.8 us | 12.6 / 12.7 / 12.6 us |

Neutral on the backward as a whole (≤ 0.7%). `compute_dh0_triton` itself pays about 0.8 us (6%) at the short-chunk shape for the extra `dht` load; that call is a few percent of the backward, so it does not show up in the totals.

The slot index could equally be written as an unclamped `start_tok = T - (W - 1)` with `w_idx = 1 + (offset - start_tok)` — the form the ascend port uses. Measured against this PR over three interleaved rounds on the varlen `initial_state + dht` backward, that variant is the same within 0.4% (485.0–485.6 us against 486.4–486.7 us here). Either form is fine; this one keeps the change to the slot index alone and leaves the mask as it was.

## Breaking changes

None. `T >= W - 1` is unchanged — the clamp never fired there and a chunk that long overwrites the whole state.

## Checklist

- [x] I have read [CONTRIBUTING.md](../CONTRIBUTING.md) and follow its conventions (code style, docstrings, commit prefixes).
- [x] I have read [AGENTS.md](../AGENTS.md) and, where my change matches its scope, the relevant skill under [.agents/skills](../.agents/skills).
- [x] This is not a minor/cosmetic-only PR (typo, formatting, style-only tweaks).
- [x] Dependent tests pass locally or in CI; new behavior is covered by tests where applicable.
- [x] Kernel changes include same-hardware before/after benchmark numbers (dense + varlen where applicable).

## Environment

- Commit base: 033b19e8 (main)
- GPU: NVIDIA B200 (sm_100), driver CUDA 13.2
- torch 2.13.0+cu130, triton 3.7.1, Python 3.12

### CODE REVIEW

**Review by zhiyuan1i — APPROVED** [review]
LGTM. In real world, there should not have a sequence shorter than W-1 Lol

**Comment by zhiyuan1i** [comment]
@sunyi0505 @zheliuyu could you help look at the A2 NPU failure on this PR? The new short-chunk backward cases (T < W - 1, with initial_state and final-state grad) fail there with `dw: NaN detected in tri`, but only the single-token + silu one; T = 2/3 and the no-activation variants pass. This PR only touches the CUDA-side Triton kernels, so this looks like a pre-existing gap in the triton-ascend port that the new shapes expose.

From a static read of `fla/modules/backends/triton_ascend/causal_conv1d.py`, two things:

The NaN: the failing call is T = 1 with initial_state and silu, which goes through `_launch_fwd_core_scalar` + `_launch_silu_bwd` and then the split backward with a degenerate BT = 1 tile. The y and dx asserts pass before dw fails, so all the kernel inputs are finite, and every load/store in the dwdb kernel is masked and in-bounds on paper. That smells like a codegen-level issue on the degenerate tile (masked-load `other=0` not honored, or the [1, N] reduce/store) rather than an indexing bug — needs an on-device repro to pin down.

Independently of the NaN: `compute_dh0_npu` takes no `dht` at all, so the final-state gradient never reaches `dh0` — the pass-through term this PR adds on the GPU side (for sequences shorter than W - 1) is structurally missing on Ascend. Even once the NaN is fixed, these cases should still fail the `dh0` assert on NPU.

No rush — real-world sequences are never this short, but it would be good to understand the NaN. Failing config: `tests/modules/test_conv.py::test_conv_cache_backward`, the new T = 1 rows (W = 4 and W = 3), silu, fp32.

**Comment by sunyi0505** [comment]
> @sunyi0505 @zheliuyu could you help look at the A2 NPU failure on this PR? The new short-chunk backward cases (T < W - 1, with initial_state and final-state grad) fail there with `dw: NaN detected in tri`, but only the single-token + silu one; T = 2/3 and the no-activation variants pass. This PR only touches the CUDA-side Triton kernels, so this looks like a pre-existing gap in the triton-ascend port that the new shapes expose.
> 
> From a static read of `fla/modules/backends/triton_ascend/causal_conv1d.py`, two things:
> 
> The NaN: the failing call is T = 1 with initial_state and silu, which goes through `_launch_fwd_core_scalar` + `_launch_silu_bwd` and then the split backward with a degenerate BT = 1 tile. The y and dx asserts pass before dw fails, so all the kernel inputs are finite, and every load/store in the dwdb kernel is masked and in-bounds on paper. That smells like a codegen-level issue on the degenerate tile (masked-load `other=0` not honored, or the [1, N] reduce/store) rather than an indexing bug — needs an on-device repro to pin down.
> 
> Independently of the NaN: `compute_dh0_npu` takes no `dht` at all, so the final-state gradient never reaches `dh0` — the pass-through term this PR adds on the GPU side (for sequences shorter than W - 1) is structurally missing on Ascend. Even once the NaN is fixed, these cases should still fail the `dh0` assert on NPU.
> 
> No rush — real-world sequences are never this short, but it would be good to understand the NaN. Failing config: `tests/modules/test_conv.py::test_conv_cache_backward`, the new T = 1 rows (W = 4 and W = 3), silu, fp32.

Certainly. I will analyze the specific causes of the UT failure and update them in the comments.
